t/05-scheduler-dependencies.t

```diff
@@ -24,13 +24,125 @@ use lib "$FindBin::Bin/lib";
 use OpenQA::Scheduler::Model::Jobs;
 use OpenQA::Constants 'WEBSOCKET_API_VERSION';
 use OpenQA::Test::Database;
+use Test::Exception;
 use Test::Mojo;
 use Test::More;
 use Test::Warnings;
 use OpenQA::WebSockets::Client;
 use OpenQA::Jobs::Constants;
 use Test::MockModule;
 
+subtest 'compute hierarchy of directly chained dependencies' => sub {
+    my %cluster_info = (
+        0 => {
+            directly_chained_children => [1, 6, 7, 8, 12],
```

Comment:
I still think this is heading the wrong way. I think the order in which the fanboys are executed needs to be defined on ISO post and mapped into job dependencies. As such I expect every job to have exact 0 or 1 fanboy in the DB and the cluster info would be 0 => [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]

I wouldn't leave that to the worker, so the webui can always report which job needs to run before job X runs.

Reply:
This "hierarchy" actually gives a fixed order (which the worker will blindly execute). The nesting is only for the error handling. The worker needs to know for which subsequent jobs a failure is crucial.

Reply:
> I think the order in which the fanboys are executed needs to be defined on ISO post and mapped into job dependencies.

This is what I intend to do. I don't see how this is related to this section of the code, though. The test you've been commenting on is for serializing the dependencies which are already in the database at this point.

Maybe I should rename the function from `_compute_directly_chained_job_hierarchy` to `_serialize_directly_chained_job_sequence`.

And generally I wouldn't treat directly chained dependencies more different from the other dependency types than required.

Reply:
ok, but will we also be able to calculate blocked_by without relying on the scheduler?

Reply:
> This is what I intend to do. I don't see how this is related to this section of the code, though. The test you've been commenting on is for serializing the dependencies which are already in the database at this point.

What concerns me mostly is that this crucial information - and only the scheduler and the worker have it and even the dependency graph can only show the hierarchy, but not the exact order. As such I feel this dependency *is* different than the others.

Reply:
Conclusion from the meeting:

* It would be desirable to to flatten the sequence already when inserting the dependencies to the database so we can also display that sequence to the user.
* However, that would discard the information to decide which jobs to skip because of previous failures and which can still run. So we keep the current approach.
* Displaying flattened direct sequences without loosing that information is also difficult to implement anyways. Therefore we keep displaying the hierarchical form for now. (I suppose the graph framework we use would require one to explicitly make cluster IDs for the "directly chained branches". So we likely needed a different function to produce the JSON for the graph framework anyways.)